include/hip/devicelib/single_precision/sp_intrinsics.hh

```diff
@@ -25,52 +25,297 @@
-static inline __device__ float __sinf(float x) { return ::sin(x); }
-static inline __device__ float __tanf(float x) { return ::tan(x); }
-
-// __device__​ float __exp10f ( float  x )
-// __device__​ float __fadd_rd ( float  x, float  y )
-// __device__​ float __fadd_rn ( float  x, float  y )
-// __device__​ float __fadd_ru ( float  x, float  y )
-// __device__​ float __fadd_rz ( float  x, float  y )
-// __device__​ float __fdiv_rd ( float  x, float  y )
-// __device__​ float __fdiv_rn ( float  x, float  y )
-// __device__​ float __fdiv_ru ( float  x, float  y )
-// __device__​ float __fdiv_rz ( float  x, float  y )
-__device__ float __fdividef(float x, float y);
-// __device__​ float __fmaf_ieee_rd ( float  x, float  y, float  z )
-// __device__​ float __fmaf_ieee_rn ( float  x, float  y, float  z )
-// __device__​ float __fmaf_ieee_ru ( float  x, float  y, float  z )
-// __device__​ float __fmaf_ieee_rz ( float  x, float  y, float  z )
-// __device__​ float __fmaf_rd ( float  x, float  y, float  z )
-// __device__​ float __fmaf_rn ( float  x, float  y, float  z )
-// __device__​ float __fmaf_ru ( float  x, float  y, float  z )
-// __device__​ float __fmaf_rz ( float  x, float  y, float  z )
-// __device__​ float __fmul_rd ( float  x, float  y )
-// __device__​ float __fmul_rn ( float  x, float  y )
-// __device__​ float __fmul_ru ( float  x, float  y )
-// __device__​ float __fmul_rz ( float  x, float  y )
-// __device__​ float __frcp_rd ( float  x )
-// __device__​ float __frcp_rn ( float  x )
-// __device__​ float __frcp_ru ( float  x )
-// __device__​ float __frcp_rz ( float  x )
-// __device__​ float __frsqrt_rn ( float  x )
-// __device__​ float __fsqrt_rd ( float  x )
-// __device__​ float __fsqrt_rn ( float  x )
-// __device__​ float __fsqrt_ru ( float  x )
-// __device__​ float __fsqrt_rz ( float  x )
-// __device__​ float __fsub_rd ( float  x, float  y )
-// __device__​ float __fsub_rn ( float  x, float  y )
-// __device__​ float __fsub_ru ( float  x, float  y )
-// __device__​ float __fsub_rz ( float  x, float  y )
-// __device__​ float __powf ( float  x, float  y )
-// __device__​ float __saturatef ( float  x )
-// __device__​ void __sincosf ( float  x, float* sptr, float* cptr )
+/**
+ * @brief Declare as extern - we state that these funcitons are implemented and
+ * will be found at link time
+ *
+ * The format is as follows:
+ * 1. Declare the external function which will be executed with the appropriate
+ * linkage. Inline comment specifying where the implementation is coming from.
+ * (OpenCL, OCML, custom) note: some of these declarations are not strictly
+ * necesary but are included for completeness.
+ * 2. If necessary, define the type specific function and bind it to the
+ * function declared in 1. cosf(x) -> cos(x)
+ */
+
+extern "C++" __device__ float native_cos(float x); // OpenCL
+extern "C++" inline __device__ float __cosf(float x) { return native_cos(x); }
```

Comment:
Unfortunately we cannot use the native_* functions for CUDA intrinsics as they do not guarantee to fulfil their minimal accuracy. https://github.com/CHIP-SPV/chip-spv/issues/222 this is likely why the blacksholes starts to fail again with this patch?

Reply:
Right, will a `CHIP_FAST_MATH` guard for this.